link.go

```diff
@@ -284,63 +328,45 @@ func (l *link) muxClose(ctx context.Context, err *Error, deferred func(), onRXTr
 		Error:  err,
 	}
 
-Loop:
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case l.session.tx <- fr:
-			// after sending the detach frame, break the read loop
-			break Loop
-		case fr := <-l.rx:
-			// read from link to avoid blocking session.mux
-			switch fr := fr.(type) {
-			case *frames.PerformDetach:
-				if fr.Closed {
-					l.detachReceived = true
-				}
-			case *frames.PerformTransfer:
-				if onRXTransfer != nil {
-					onRXTransfer(*fr)
-				}
-			}
-		case <-l.session.done:
-			if l.doneErr == nil {
-				l.doneErr = l.session.doneErr
-			}
-			return
+	select {
+	case <-ctx.Done():
+		return
+	case l.session.tx <- fr:
+		// frame sent to our session mux
+	case <-l.session.done:
+		if l.doneErr == nil {
+			l.doneErr = l.session.doneErr
 		}
+		return
 	}
 
-	// don't wait for remote to detach when already received
+	// if the peer initiated the close then we just sent the ack so we're done
 	if l.detachReceived {
 		return
 	}
 
+	// wait for the ack
 	for {
-		select {
-		case <-ctx.Done():
+		fr, err := l.waitForFrame(ctx)
+		if isContextErr(err) {
+			// TODO: expired context
 			return
```

Comment:
(possibly for future issue)
So at this point we'e sent the DETACH frame but got no response. Should this be signalled with an error?

Reply:
Will fix as part of https://github.com/Azure/go-amqp/issues/246